core/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -240,61 +261,83 @@ public void beforeListVariableChanged(ListVariableDescriptor<Solution_> variable
     public void afterListVariableChanged(ListVariableDescriptor<Solution_> variableDescriptor, Object entity, int fromIndex,
             int toIndex) {
         var notifiables = notifiableRegistry.get(variableDescriptor);
+        var notification = Notification.<Solution_> listVariableChanged(entity, fromIndex, toIndex);
         if (!notifiables.isEmpty()) {
-            ListVariableNotification<Solution_> notification = Notification.listVariableChanged(entity, fromIndex, toIndex);
             for (var notifiable : notifiables) {
                 notifiable.notifyAfter(notification);
             }
             notificationQueuesAreEmpty = false;
         }
-        listVariableEventList.add(new ListVariableEvent(entity, fromIndex, toIndex));
+        if (!cascadingUpdateShadowVarDescriptorList.isEmpty()) { // Only necessary if there is a cascade.
+            listVariableChangedNotificationList.add(notification);
+        }
```

Comment:
No need to create the event if we already have the notification.
Help the GC do less work, help yourself.